src/DotNetWorker.ApplicationInsights/FunctionsTelemetryModule.cs

```diff
@@ -98,10 +102,16 @@ internal static void TrackExceptionTelemetryFromActivityEvent(ActivityEvent acti
             }
         }
 
-        public void Dispose()
+        public async ValueTask DisposeAsync()
         {
-            _telemetryClient?.Flush();
             _listener?.Dispose();
+
+            if (_telemetryClient is not null)
+            {
+                await _telemetryClient.FlushAsync(CancellationToken.None);
```

Comment:
FlushAsync may take longer, especially if the telemetry client is in backoff retry mode. It would be safe to cancel flush after few seconds if it's not complete.

Reply:
Agree. We're working through getting this right in the host as well -- at some point we just give up -- but that's not even enough time in some cases and we end up with CRIs with missing telemetry during a shutdown (I think we give up after 2 seconds there).

I could start with 5 second cancelation and let people configure it? I was noticing that even AppInsights doesn't have great handling for this: https://github.com/microsoft/ApplicationInsights-dotnet/blob/main/LOGGING/src/ILogger/ApplicationInsightsLoggerProvider.cs#L111-L113

Reply:
Michael has PR open to improve this on host. https://github.com/Azure/azure-webjobs-sdk/pull/2939
He is logging if the flush fail to complete in stipulated time.